src/Microsoft.Extensions.WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,391 @@
+                if (masked)
+                {
+                    var maskingKeyStart = buffer.Start;
+                    maskingKey = buffer.Slice(0, 4);
+
+                    // Mark the masking key examined but not consumed
+                    _inbound.Advance(maskingKeyStart, buffer.Start);
+                }
+                else
+                {
+                    // Mark everything consumed, there was no masking key
+                    _inbound.Advance(buffer.Start);
+                }
+            }
+
+            var payload = default(ReadableBuffer);
+            if (payloadLen > 0)
+            {
+                buffer = await _inbound.ReadAtLeastAsync(payloadLen, cancellationToken);
+
+                if(masked)
+                {
+                    // Need to skip the masking key, since we didn't consume it.
+                    buffer = buffer.Slice(4);
+                }
+                payload = buffer.Slice(0, payloadLen);
+
+                // We don't slice the buffer because we actually want to keep the payload marked unconsumed until the next read
+                // Instead, we capture the position after the payload as the "lastFrameCursor"
+                _lastPayloadEnd = buffer.Slice(payloadLen).Start;
+
+                if (masked)
+                {
+                    unsafe
+                    {
+                        // TODO: Unmask in place? Can we do that safely? If so we can avoid a heap allocation here.
+                        var unmasked = new byte[payload.Length];
+                        payload.CopyTo(unmasked);
+
+                        // Get the masking key into a Span<byte> as efficiently as possible
+                        var maskingKeySpan = Span<byte>.Empty;
+
+                        // Stack space for storing the masking key if needed.
+                        var maskingKeyStorage = stackalloc byte[4];
+
+                        if (maskingKey.IsSingleSpan)
+                        {
+                            // It's already a single span!
+                            maskingKeySpan = maskingKey.First.Span;
+                        }
+                        else
+                        {
+                            // We need to copy it to a single span.
+                            maskingKeySpan = new Span<byte>(maskingKeyStorage, 4);
+                            maskingKey.CopyTo(maskingKeySpan);
+                        }
+
+                        // Either we initialized it with a length, or it came from the maskingKey buffer which should already have been sliced.
+                        Debug.Assert(maskingKeySpan.Length == 4);
+
+                        // Unmask
+                        for (int i = 0; i < payload.Length; i++)
+                        {
+                            unmasked[i] ^= maskingKeySpan[i % 4];
+                        }
+                        payload = ReadableBuffer.Create(unmasked, 0, unmasked.Length);
```

Comment:
There's an integration piece here that hasn't been done (this is prototype code from a different scenario being merged in) where we may just end up writing the payload directly into a WritableChannel for the "Socket" connection to use, so this code will be seeing a lot of churn shortly.